### PR TITLE
Entity system

### PR DESCRIPTION
**Pull request by Trompettesib**

### CODE REVIEW

File: cubic-server/Dimension.hpp [hunk 1]
```diff
@@ -19,8 +23,12 @@ class Dimension
     virtual void initialize();
     virtual void tick();
     virtual World *getWorld() const;
+    virtual std::vector<Entity *> getEntities();
+    virtual void forEachEntity(std::function<void(Entity *)> callback);
+    virtual void forEachEntityIf(std::function<bool(Entity *)> predicate, std::function<void(Entity *)> callback);
```
Comment (huntears) [inline cubic-server/Dimension.hpp:28]: Please put the callback first

File: cubic-server/Dimension.hpp [hunk 2]
```diff
@@ -19,8 +23,12 @@ class Dimension
     virtual void initialize();
     virtual void tick();
     virtual World *getWorld() const;
+    virtual std::vector<Entity *> getEntities();
```
Comment (huntears) [inline cubic-server/Dimension.hpp:26]: You sure you want a copy here ?

File: cubic-server/Entity.hpp [hunk 3]
```diff
@@ -4,18 +4,53 @@
...
+        SNEAKING,
+        LONG_JUMPING,
+        DYING,
+        CROAKING,
+        USING_TONGUE,
+        ROARING,
+        SNIFFING,
+        EMERGING,
+        DIGGING
+    };
```
Comment (huntears) [inline cubic-server/Entity.hpp:27]: Please don't put then in uppercase like this.
Just use PascalCase

File: cubic-server/Entity.cpp [hunk 4]
```diff
@@ -1,5 +1,27 @@
...
+    _airTicks = 300;
+    _customName = "";
+    _customNameVisible = false;
+    _silent = false;
+    _noGravity = false;
+    _pose = Pose::STANDING;
+    _tickFrozenInPowderedSnow = 0;
+}
+
+Entity::~Entity()
```
Comment (huntears) [inline cubic-server/Entity.cpp:21]: The destructor has been deleted

File: cubic-server/World.cpp [hunk 5]
```diff
@@ -32,5 +32,21 @@ std::shared_ptr<Dimension> World::getDimension(const std::string_view &name) con
 
 std::vector<Entity *> World::getEntities()
 {
-    return _entities;
+    std::vector<Entity *> entities;
+    for (auto & _dimension : _dimensions)
+        for (auto & _entity : _dimension.second->getEntities())
+            entities.push_back(_entity);
+    return entities;
 }
```
Comment (huntears) [inline cubic-server/World.cpp:40]: Caching that would be really nice.

File: cubic-server/Entity.cpp (hunk 4 above)
Comment (Trompettesib) [reply cubic-server/Entity.cpp:21]: I know but if I remove that the linker cannot link for a reason I don't understand
image

Comment (STMiki) [reply cubic-server/Entity.cpp:21]: https://stackoverflow.com/a/1219618

> Note: The destructor is the only method that even if it is pure virtual has to have an implementation in order to instantiate derived classes (yes pure virtual functions can have implementations).

**Review by huntears — APPROVED** [review]

**Review by STMiki — APPROVED** [review]
👍

**Pull request merged**